### PR TITLE
[Exporter Refactor] Base Exporter class and example implementations

### PR DESCRIPTION
**Pull request by corey-nm**
Remaining todos:
- [x] Add tests for SkipInputQuantize
- [x] Add tests for TorchToONNX
- [x] Add tests for ONNXToDeepsarse
- [x] Add regression tests

### CODE REVIEW

File: src/sparseml/exporters/torch_to_onnx.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+            raise TypeError(
+                f"Expected torch.nn.Module, found {type(pre_transforms_model)}"
+            )
+        post_transforms_model: onnx.ModelProto = self.apply(pre_transforms_model)
+        # sanity check
+        assert isinstance(post_transforms_model, onnx.ModelProto)
+        onnx.save(post_transforms_model, file_path)
+
+
+class _TorchOnnxExport(BaseTransform):
```
Comment (corey-nm) [inline src/sparseml/exporters/torch_to_onnx.py:91]: Wasn't sure if this should go under `transforms/`

File: src/sparseml/exporters/torch_to_onnx.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+        self.export_kwargs = export_kwargs or {}
+
+    def pre_validate(self, model: Any) -> torch.nn.Module:
+        if not isinstance(model, torch.nn.Module):
+            raise TypeError(f"Expected torch.nn.Module, found {type(model)}")
+        return model
+
+    def post_validate(self, model: Any) -> onnx.ModelProto:
+        # sanity check
+        assert isinstance(model, onnx.ModelProto)
```
Comment (corey-nm) [inline src/sparseml/exporters/torch_to_onnx.py:127]:
```suggestion
        assert isinstance(model, onnx.ModelProto)
        onnx.checker.check_model(model)
```

File: src/sparseml/exporters/onnx_to_deepsparse.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        use_qlinear_conv: bool = False,
+        skip_input_quantize: bool = False,
+    ) -> None:
+        conv_transform = (
+            sparseml_transforms.ConvToQLinearConv()
+            if use_qlinear_conv
+            else sparseml_transforms.ConvertQuantizableConvInteger()
+        )
+
+        transforms = [
```
Comment (bfineran) [inline src/sparseml/exporters/onnx_to_deepsparse.py:60]: let's think about splitting this up into the first half and the actual Q/DQ conversions - first half is mostly around cleaning up the torch export graph

File: src/sparseml/exporters/torch_to_onnx.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+# limitations under the License.
+
+import collections
+import tempfile
+import warnings
+from copy import deepcopy
+from typing import Any, Dict, Iterable, List
+
+import onnx
+import torch
```
[the remark this answers is not in the document]
Comment (bfineran) [reply src/sparseml/exporters/torch_to_onnx.py:22]: yup

File: src/sparseml/exporters/onnx_to_deepsparse.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        return model if self.inplace else deepcopy(model)
+
+    def post_validate(self, model: Any) -> onnx.ModelProto:
+        # sanity check
+        if not isinstance(model, onnx.ModelProto):
+            raise TypeError(f"Expected onnx.ModelProto, found {type(model)}")
+        return model
+
+    def export(self, pre_transforms_model: onnx.ModelProto, file_path: str):
+        if self.export_input_model:
```
Comment (corey-nm) [inline src/sparseml/exporters/onnx_to_deepsparse.py:107]: Can also check environment variable for this

File: src/sparseml/exporters/onnx_to_deepsparse.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+
+class ONNXToDeepsparse(BaseExporter):
+    """
+    Optimizes an `onnx.ModelProto` for the deepsparse engine by applying a
+    series of transformations to a onnx graph with quantize operations.
+
+    Usage:
+    ```python
+    model: onnx.ModelProto = ... # could be from TorchToOnnx() or somewhere else
```
Comment (dbogunowicz) [inline src/sparseml/exporters/onnx_to_deepsparse.py:32]:
```suggestion
    model: onnx.ModelProto = ... # could be a model retrieved previously from TorchToOnnx() or somewhere else
```

File: src/sparseml/exporters/onnx_to_deepsparse.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    def post_validate(self, model: Any) -> onnx.ModelProto:
+        # sanity check
+        if not isinstance(model, onnx.ModelProto):
+            raise TypeError(f"Expected onnx.ModelProto, found {type(model)}")
+        return model
+
+    def export(self, pre_transforms_model: onnx.ModelProto, file_path: str):
+        if self.export_input_model:
+            onnx.save(
+                pre_transforms_model, file_path.replace(".onnx", ".pre-deepsparse.onnx")
```
Comment (dbogunowicz) [inline src/sparseml/exporters/onnx_to_deepsparse.py:109]:
```suggestion
                pre_transforms_model, file_path.replace(".onnx", "._deepsparse.onnx")
```

File: src/sparseml/pytorch/torch_to_onnx_exporter.py [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+                    sample_batch, opset, disable_bn_fusing, **export_kwargs
+                ),
+                sparseml_transforms.FoldIdentityInitializers(),
+                sparseml_transforms.FlattenQParams(),
+                sparseml_transforms.UnwrapBatchNorms(),
+                sparseml_transforms.DeleteTrivialOnnxAdds(),
+            ],
+        )
+
+    def pre_validate(self, module: Any) -> torch.nn.Module:
```
Comment (dbogunowicz) [inline src/sparseml/pytorch/torch_to_onnx_exporter.py:85]:
```suggestion
    def pre_validate(self, module: torch.nn.Module) -> torch.nn.Module:
```

File: src/sparseml/pytorch/torch_to_onnx_exporter.py [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+        if not isinstance(module, torch.nn.Module):
+            raise TypeError(f"Expected torch.nn.Module, found {type(module)}")
+
+        if is_parallel_model(module):
+            module = module.module
+
+        return deepcopy(module).to("cpu").eval()
+
+    def post_validate(self, model: Any) -> onnx.ModelProto:
```
Comment (dbogunowicz) [inline src/sparseml/pytorch/torch_to_onnx_exporter.py:94]:
```suggestion
    def post_validate(self, model:  torch.nn.Module) -> onnx.ModelProto:
```

File: src/sparseml/pytorch/torch_to_onnx_exporter.py [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+                "Consider using OrderedDict for this purpose.",
+                UserWarning,
+            )
+
+        self.sample_batch = sample_batch
+        self.opset = opset
+        self.disable_bn_fusing = disable_bn_fusing
+        self.export_kwargs = export_kwargs or {}
+
+    def pre_validate(self, model: Any) -> torch.nn.Module:
```
Comment (dbogunowicz) [inline src/sparseml/pytorch/torch_to_onnx_exporter.py:136]:
```suggestion
    def pre_validate(self, model: torch.nn.Module) -> torch.nn.Module:
```

File: src/sparseml/pytorch/torch_to_onnx_exporter.py [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+        if not isinstance(model, torch.nn.Module):
+            raise TypeError(f"Expected torch.nn.Module, found {type(model)}")
+        return model
+
+    def post_validate(self, model: Any) -> onnx.ModelProto:
```
Comment (dbogunowicz) [inline src/sparseml/pytorch/torch_to_onnx_exporter.py:141]:
```suggestion
    def post_validate(self, model: onnx.ModelProto) -> onnx.ModelProto:
```

File: src/sparseml/pytorch/torch_to_onnx_exporter.py [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+        opset: int = TORCH_DEFAULT_ONNX_OPSET,
+        disable_bn_fusing: bool = True,
+        **export_kwargs,
+    ):
+        super().__init__(
+            [
+                _TorchOnnxExport(
+                    sample_batch, opset, disable_bn_fusing, **export_kwargs
+                ),
+                sparseml_transforms.FoldIdentityInitializers(),
```
Comment (dbogunowicz) [inline src/sparseml/pytorch/torch_to_onnx_exporter.py:78]: This is where we start working with the onnx model right? Let's maybe mention it in the docstrings/comments.

**Review by dbogunowicz — APPROVED** [review]
Looking good!

File: src/sparseml/exporters/onnx_to_deepsparse.py (hunk 7 above)
Comment (corey-nm) [reply src/sparseml/exporters/onnx_to_deepsparse.py:110]: Hmm what about `.preqat.onnx`?

File: src/sparseml/pytorch/torch_to_onnx_exporter.py (hunk 9 above)
Comment (corey-nm) [reply src/sparseml/pytorch/torch_to_onnx_exporter.py:94]: I think this should be onnx.ModelProto - I added that

File: src/sparseml/exporters/transforms/delete_repeated_qdq.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
```
Comment (corey-nm) [inline src/sparseml/exporters/transforms/delete_repeated_qdq.py:1]: Review of this was moved to #1257

File: src/sparseml/exporters/transforms/skip_input_quantize.py [hunk 14]
```diff
@@ -0,0 +1,81 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
```
Comment (corey-nm) [inline src/sparseml/exporters/transforms/skip_input_quantize.py:1]: Review of this was moved to #1256

**Review by bfineran — APPROVED** [review]
LGTM, small nit may be adding `Exporter` to all exporter class names

**Pull request merged**